File: BinarySearchTree/Check BST/checkBST.cpp
#include <iostream>
#include <queue>
using namespace std;
template <typename T> class BSTNode
{
public:
	BSTNode<T> *left;
	BSTNode<T> *right;
	T data;
	BSTNode(T data)
	{
		this->data=data;
	}
};
class Answer
{
public:
	int minValue;
	int maxValue;
	bool isBst;
	Answer(int minValue , int maxValue , bool isBst)
	{
		this->minValue=minValue;
		this->maxValue=maxValue;
		this->isBst=isBst;
	}
};
BSTNode<int> * takeInput()
{
	int data;
	cin>>data;
	if (data==-1)	return NULL;
	//otherwise
	BSTNode<int> *root=new BSTNode<int>(data);
	queue<BSTNode<int>*> pending;
	pending.push(root);
	while(pending.size())
	{
		BSTNode<int> *temp=pending.front();
		pending.pop();
		//left subtree'r root node
		cin>>data;
		if (data!=-1)
		{
			temp->left=new BSTNode<int>(data);
			pending.push(temp->left);
		}else{
			temp->left=NULL;
		}
		//right subtree'r root node
		cin>>data;
		if (data!=-1)
		{
			temp->right=new BSTNode<int>(data);
			pending.push(temp->right);
		}else{
			temp->right=NULL;
		}
	}
	return root;
}
Answer * checkBst(BSTNode<int> *root)
{
	if (!root)
	{
		Answer *ans=new Answer(INT_MAX , INT_MIN , true);
		return ans;
	}
	Answer * leftAns=checkBst(root->left);
	Answer * rightAns=checkBst(root->right);
	int minValue=min(root->data , min(leftAns->minValue , rightAns->minValue));
	int maxValue=max(root->data , max(leftAns->maxValue , rightAns->maxValue));
	bool is_bst=true;
	//there is total four condition
	if ((leftAns->maxValue)>(root->data))
	{
		is_bst=false;
	}
	if ((rightAns->minValue)<(root->data))
	{
		is_bst=false;
	}
	if (!(leftAns->isBst) || ! (rightAns->isBst))
	{
		is_bst=false;
	}
	Answer *ans=new Answer(minValue , maxValue , is_bst);
	return ans;
}
int main()
{
	BSTNode<int>*root=takeInput();
	Answer *ans=checkBst(root);
	if (ans->isBst)
	{
		cout<<"true";
	}else{
		cout<<"false";
	}
	cout<<endl;
	cout<<ans->minValue<<" "<<ans->maxValue;
}

File: BinarySearchTree/Check BST/CheckBST.java
import java.util.Scanner;
import java.util.Queue;
import java.util.LinkedList;
class BSTNode<T>
{
	public BSTNode<T> left;
	public BSTNode<T> right;
	public T data;
	BSTNode(T data)
	{
		this.data=data;
	}
};
class CheckBST
{
	private static Scanner scan=new Scanner(System.in);
	private static BSTNode<Integer> takeInput()
	{
		int data=scan.nextInt();
		if (data==-1) {
			return null;
		}
		BSTNode<Integer> root=new BSTNode<Integer>(data);
		Queue<BSTNode<Integer>> pending=new LinkedList<BSTNode<Integer>>();
		pending.add(root);
		while(!pending.isEmpty())
		{
			BSTNode<Integer> temp=pending.remove();
			data=scan.nextInt();
			if (data!=-1) {
				temp.left=new BSTNode<Integer>(data);
				pending.add(temp.left);
			}else{
				temp.left=null;
			}
			data=scan.nextInt();
			if (data!=-1) {
				temp.right=new BSTNode<Integer>(data);
				pending.add(temp.right);
			}else{
				temp.right=null;
			}
		}
		return root;
	}
	private static int findMin(BSTNode<Integer> root)
	{
		if (root==null) {
			return Integer.MAX_VALUE;
		}
		int leftMin=findMin(root.left);
		int rightMin=findMin(root.right);
		return Math.min(root.data , Math.min(leftMin , rightMin));
	}
	private static int findMax(BSTNode<Integer> root)
	{
		if (root==null) {
			return Integer.MIN_VALUE;
		}
		int leftMax=findMax(root.left);
		int rightMAx=findMax(root.right);
		return Math.max(root.data , Math.max(leftMax , rightMAx));
	}
	private static boolean isBSt(BSTNode<Integer> root)
	{
		if (root==null) {
			return true;
		}
		int min=findMin(root.right);
		int max=findMax(root.left);
		if (root.data>min) {
			return false;
		}
		if (root.data<=max) {
			return false;
		}
		boolean leftAns=isBSt(root.left);
		boolean rightAns=isBSt(root.right);
		return (leftAns && rightAns);
	}
	public static void main(String[] args) {
		BSTNode<Integer> root=takeInput();
		boolean isBSt=isBSt(root);
		System.out.println(isBSt);
	}
}